src/coreclr/jit/lowerxarch.cpp

```diff
@@ -8331,6 +8403,51 @@ void Lowering::ContainCheckHWIntrinsic(GenTreeHWIntrinsic* node)
                             break;
                         }
 
+                        case NI_AVX512F_TernaryLogic:
+                        case NI_AVX512F_VL_TernaryLogic:
+                        {
+                            if (!isContainedImm)
+                            {
+                                // Don't contain if we're generating a jmp table fallback
+                                break;
+                            }
+
+                            if (IsContainableHWIntrinsicOp(node, op3, &supportsRegOptional))
+                            {
+                                MakeSrcContained(node, op3);
+                            }
+                            else if (supportsRegOptional)
+                            {
+                                MakeSrcRegOptional(node, op3);
+                            }
+
+                            uint8_t                 control  = static_cast<uint8_t>(op4->AsIntCon()->gtIconVal);
+                            const TernaryLogicInfo& info     = TernaryLogicInfo::lookup(control);
+                            TernaryLogicUseFlags    useFlags = info.GetAllUseFlags();
+
+                            if (useFlags != TernaryLogicUseFlags::ABC)
```

Comment:
I am not sure I completely follow what happens when `useFlag == ABC`. Can you elaborate?

Reply:
When all three operands are used we have a real ternary instruction and don't have an optimizations or normalization to do. This is because the instruction is already "complete" (aside from constant folding) and can't represent any more state.

For the cases when all operands aren't used, then we have the ability to represent the operation more optimally. This happens by either converting it to the regular unary/binary node (e.g. `gtNewSimdBinOpNode(GT_AND, ...)`) or by normalizing the format so that the used operands are `B/C` (which allows us to not be marked `RMW` and allows greatest chance for containment). It will also simplify the logic that will be added in a future PR to combine unary/binary ops into ternary logic nodes.